Workspace inactivity notifier

davidreneuw opened the pull request:
Implemented the following:

- **Project inactivity scheduler function**: goes through all projects daily and schedules them to get their inactivity info updated using the inactivity updater function.
- **Project inactivity updater function**: for each queue element, will calculate days since the most recent login date of any user in the given workspace. Using the new project inactivity service, will interact with the `Project_Inactivity` db table to update the `DaysSinceLastLogin` of a project. Then it will send the queue item to the inactivity notifier function.
- **Project inactivity notifier function**: for each project, will check if a notification is required, and will notify all users of a workspace if it is required. A list of notification days are provided in `appsettings` (i.e. `23,28,30`). The function will send emails for all notification days once `DaysSinceLastLogin` reaches the notification day, except for the last "notification" day (in this case, 30), where instead of sending the email, it will start the deletion process (this is not implemented yet). If a workspace is `Whitelisted`, it will not be affected by this, though if a `RetirementDate` for a workspace is provided and that date is today, then the deletion process will also be started.
- **Project inactivity service**: allows interacting with the corresponding table. Implemented the necessary function so that we can later on set retirement dates and whitelisting for workspaces using a UI.

_**Notes**_: 

1. I haven't tested any of this yet and am mostly looking for conceptual/logical feedback
2. When implementing the new `Project_Inactivity` table in `DatahubProjectDBContext`, I did not understand how to properly set up the table in the `OnModelCreating` function. I will need help with that
3. This was mostly based on the existing work Iroel has done with project usage notifying

davidreneuw committed:
Big commit. Developed workspace inactivity notifier.

davidreneuw committed:
Implement whitelisting and retirement date

ErikApption commented on Portal/src/Datahub.Core/Model/Projects/Project_Inactivity.cs:
@yjmrobert Do you think we should put whitelisted on the project object?

yjmrobert replied:
We'd considered that, as well as just wrapping it in a project extension method that would get the inactivity bool if the `PortalUser`s were loaded.

That table is getting a huge mish-mash of random fields and I'd like to clean it up before production release. I'm good with either sticking it on the project object to clean up later though

davidreneuw committed:
Moving inactivity info to project object

davidreneuw committed:
Implemented new logic where only expiration date is provided. Checks that against the project's last login date or last updated date (if null). Last updated date and last updated userId now both update on user permission changes in a workspace.

davidreneuw committed:
Adding deletion capability to ResourceMessagingService.cs

davidreneuw committed:
Re-fixed minor translation issue. Fixed ProjectUserManagementServiceTests

davidreneuw committed:
Added migration for new project inactivity table and new column in project object

davidreneuw committed:
Added workspace deletion button and dialog

davidreneuw committed:
UnitTested ProjectInactivityNotifier.cs, created a new project in doing so.
Separated parameters into a provider to be able to mock these parameters for testing.
Added HasCostRecovery bool to Projects object.
Refactored ExpiryDate to OperationalWindow.

davidreneuw committed:
Removed old migration and recreated new migration to match the recent changes.

davidreneuw committed:
Refactored to fit Pascal naming scheme
Added navigation and foreign key between project object and inactivity notifications
Added IsDeleted boolean to project object
Removed old migration, recreated new migration
Updated database

davidreneuw committed:
Updated ResourceMessagingService.cs to update Deleted_DT on projects it runs SendToTerraformDeleteQueue on
Added various AsNoTracking to Linq queries when possible
Added unit test to verify correct email is generated in ProjectInactivityNotifier.cs
Added inactivity alert email template as content

davidreneuw committed:
Fixing merge conflicts

davidreneuw committed:
Fixing merge conflicts, 2nd attempts

davidreneuw commented:
Note: THIS REQUIRES A CHANGE TO `local.appsettings.json` IN `Datahub.Functions` (new queue names) AS WELL AS `appsettings.json` IN `Datahub.Portal` (delete queue name)

yjmrobert committed:
🔧 merge
